core/ledger/kvledger/txmgmt/privacyenabledstate/namespace_provider.go

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package privacyenabledstate
+
+import (
+	"github.com/hyperledger/fabric-protos-go/ledger/queryresult"
+	commonledger "github.com/hyperledger/fabric/common/ledger"
+	"github.com/hyperledger/fabric/core/ledger"
+	"github.com/hyperledger/fabric/core/ledger/kvledger/txmgmt/statedb"
+	"github.com/pkg/errors"
+)
+
+// NamespaceProvider implements statedb.NameNamespaceProvider interface
+type NamespaceProvider struct {
+	channelInfoProvider
+	ledger.DeployedChaincodeInfoProvider
+}
+
+// channelInfoProvider interface enables ledger components to retrieve channel config data in all the
+// config blocks (current and previous config blocks).
+type channelInfoProvider interface {
+	// GetAllMSPIDs retrieves the MSPIDs of application organizations in all the channel configurations,
+	// including current and previous channel configurations.
+	GetAllMSPIDs() ([]string, error)
+}
+
+// NewNamespaceProvider creates a new instance of NamespaceProvider
+func NewNamespaceProvider(channelInfoProvider channelInfoProvider, deployedccInfoProvider ledger.DeployedChaincodeInfoProvider) *NamespaceProvider {
+	return &NamespaceProvider{channelInfoProvider, deployedccInfoProvider}
+}
+
+// GetStateDBNamespaces returns the namespaces for a channel. In ledger, a private data namespace is created
+// only if the peer is a member of the collection or owns the implicit collection. However, this function
+// adopts a simple implementation that always adds private data namespace for a collection without checking
+// peer membership/ownership. As a result, it may return more private data namespaces that what should be.
+// However, it will not cause any inconsistent issue because the caller in statecouchdb will check if any
+// existing database matches the namespace and filter out all extra namespaces since no databases match them.
+// Checking peer membership is complicated because it requires retrieving all the collection configs from
+// the collection config store. Because this is a temporary function needed to retroactively build namespaces
+// when upgrading v2.0/2.1 peers to a newer v2.x version and because returning extra private data namespaces
+// does not cause inconsistence, it makes sense to use the simple implementation.
+func (p *NamespaceProvider) GetStateDBNamespaces(channelName string, vdb statedb.VersionedDB) ([]string, error) {
```

Comment:
In the connection with the comment in `kv_ledger_provider`, after looking at this code here, I feel that the logic in this funciton better suits in the package `kvledger`. That would prevent this high level logic to percolate much deeper in the code - wdyt?

Reply:
Due to the dependency on `deriveHashedDataNs()` and `derivePvtDataNs()`, it may be bit challenging to move the whole `GetStateDBNamespaces()` to `kvledger` pkg.

Reply:
This function has logic (e.g. derivePvtDataNs) specific to the `privacyenabledstate` package, therefore cannot be moved to kvledger. Will split it into two functions: one is moved to channelInfoProvider in the `kvledger` package and the other stays in `privacyenabledstate` package.